File: experiments/03_synth_data_generation/py_src/batch02_gcd_math.py
"""Batch 02: Number theory and math functions — pure computation."""

def gcd(a: int, b: int) -> int:
    while b != 0:
        a, b = b, a % b
    return a

def lcm(a: int, b: int) -> int:
    if a == 0 or b == 0:
        return 0
    return (a // gcd(a, b)) * b

def power(base: int, exp: int) -> int:
    result = 1
    while exp > 0:
        if exp % 2 == 1:
            result *= base
        base *= base
        exp //= 2
    return result

def is_prime(n: int) -> bool:
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    return True

def mod_exp(base: int, exp: int, mod: int) -> int:
    if mod == 0:
        return 0
    result = 1
    base = base % mod
    while exp > 0:
        if exp % 2 == 1:
            result = (result * base) % mod
        exp //= 2
        base = (base * base) % mod
    return result

def sum_digits(n: int) -> int:
    s = 0
    while n > 0:
        s += n % 10
        n //= 10
    return s

def count_digits(n: int) -> int:
    if n == 0:
        return 1
    c = 0
    while n > 0:
        c += 1
        n //= 10
    return c

File: experiments/03_synth_data_generation/c_src/batch02_gcd_math.c
/* Batch 02: Number theory and math functions — pure computation */

unsigned int gcd(unsigned int a, unsigned int b) {
    while (b != 0) {
        unsigned int t = b;
        b = a % b;
        a = t;
    }
    return a;
}

unsigned int lcm(unsigned int a, unsigned int b) {
    if (a == 0 || b == 0) return 0;
    return (a / gcd(a, b)) * b;
}

unsigned int power(unsigned int base, unsigned int exp) {
    unsigned int result = 1;
    while (exp > 0) {
        if (exp % 2 == 1) {
            result *= base;
        }
        base *= base;
        exp /= 2;
    }
    return result;
}

int is_prime(unsigned int n) {
    if (n < 2) return 0;
    if (n == 2) return 1;
    if (n % 2 == 0) return 0;
    for (unsigned int i = 3; i * i <= n; i += 2) {
        if (n % i == 0) return 0;
    }
    return 1;
}

unsigned int mod_exp(unsigned int base, unsigned int exp, unsigned int mod) {
    if (mod == 0) return 0;
    unsigned int result = 1;
    base = base % mod;
    while (exp > 0) {
        if (exp % 2 == 1) {
            result = (result * base) % mod;
        }
        exp /= 2;
        base = (base * base) % mod;
    }
    return result;
}

unsigned int sum_digits(unsigned int n) {
    unsigned int s = 0;
    while (n > 0) {
        s += n % 10;
        n /= 10;
    }
    return s;
}

unsigned int count_digits(unsigned int n) {
    if (n == 0) return 1;
    unsigned int c = 0;
    while (n > 0) {
        c++;
        n /= 10;
    }
    return c;
}
